fix(ast): include `directives` in `body`

hi-ogawa opened the pull request:
Part of https://github.com/oxc-project/oxc/issues/2854
Related: https://github.com/oxc-project/oxc/pull/8972#discussion_r1947996248

I updated `append_to` macro to support concatenating two `Vec`. The logic is very similar so I reused it, but maybe it's better to introduce a new macro or maybe rename `append_to`?

hi-ogawa commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8981?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#8981** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8981?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8981?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#8975** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8975?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#8974** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8974?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#8973** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8973?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#8972** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8972?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#8976** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8976?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8981).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented on tasks/ast_tools/src/derives/estree.rs:
```diff
@@ -299,10 +299,16 @@ impl<'s> StructSerializerGenerator<'s> {
-                AppendTo { array: &#value, after: &#self_path.#append_from_ident }
+            let append_field = &struct_def.fields[append_field_index];
+            let append_from_ident = append_field.ident();
+            let wrapper = if append_field.type_def(self.schema).is_option() {
+                quote! { AppendTo }
+            } else {
+                quote! { oxc_estree::ser::AppendToConcat }
```
Style nit: Can we make `AppendTo` and `AppendToConcat` consistent? Either (1) add import of  `AppendToConcat` in `fn prelude` (same as `AppendTo`), or (2) remove `AppendTo` from the prelude, and include both with their full path `oxc_estree::ser::...` here.

overlookmotel commented on tasks/ast_tools/src/generators/typescript.rs:
```diff
@@ -170,8 +170,18 @@ fn generate_ts_type_def_for_struct_field<'s>(
         let appended_field = struct_def.fields[append_field_index].type_def(schema);
-        let appended_field = appended_field.as_option().unwrap();
-        let appended_type_name = ts_type_name(appended_field.inner_type(schema), schema);
+        let appended_field = match appended_field.as_option() {
+            Some(option) => option.inner_type(schema),
+            None => match appended_field.as_vec() {
+                Some(vec) => vec.inner_type(schema),
+                None => panic!(
+                    "Appended field must be `Option<T>` or `Vec<T>`: `{}::{}`",
+                    struct_def.name(),
+                    appended_field.name()
+                ),
+            },
+        };
```
I think this is a bit clearer:

```suggestion
        let appended_field = match appended_field {
            TypeDef::Option(option_def) => option_def.inner_type(schema),
            TypeDef::Vec(vec_def) => vec_def.inner_type(schema),
            _ => panic!(
                "Appended field must be `Option<T>` or `Vec<T>`: `{}::{}`",
                struct_def.name(),
                appended_field.name()
            ),
        };
```

overlookmotel replied:
Actually, `appended_field.name()` in the `panic!` should be the field name, not the name of the *type* of the field. So we probably need to call the type `appended_field_type`, so that `appended_field` is still available here as a reference to the `FieldDef`.

overlookmotel commented:
### Merge activity

* **Feb 10, 8:44 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

hi-ogawa committed:
fix(ast): include `directives` in `body` (#8981)

Part of https://github.com/oxc-project/oxc/issues/2854
Related: https://github.com/oxc-project/oxc/pull/8972#discussion_r1947996248

I updated `append_to` macro to support concatenating two `Vec`. The logic is very similar so I reused it, but maybe it's better to introduce a new macro or maybe rename `append_to`?
